File: kafka-test/producer.py
import os
from confluent_kafka import Producer
from confluent_kafka.admin import AdminClient, NewTopic

# Kafka configuration
kafka_config = {
    'bootstrap.server': os.getenv('KAFKA_BOOTSTRAP_SERVERS', 'localhost:9092')
}
admin_client = AdminClient({'bootstrap.servers': kafka_config['bootstrap.servers']})

print(f"Kafka configuration: {kafka_config}")
print(f"bootstrap.servers: {os.getenv('KAFKA_BOOTSTRAP_SERVERS', 'localhost:9092')}")

def create_producer():
    return Producer(kafka_config)

def delivery_report(err, msg):
    """ Called once for each message produced to indicate delivery result. """
    if err is not None:
        print(f"Message delivery failed: {err}")
    else:
        print(f"Message delivered to {msg.topic()} [{msg.partition()}]")

def topic_exists(topic_name):
    return topic_name in admin_client.list_topics().topics

    # Create a Kafka topic if it doesn't exist
def create_topic( topic_name):
    if not topic_exists(topic_name):
        new_topic = NewTopic(topic_name, num_partitions=1, replication_factor=1)
        admin_client.create_topics([new_topic])


def main():
    # Create a Kafka producer
    producer = create_producer()

    while True:
        # Take topic name from user input
        topic = input("Enter the Kafka topic name (or type 'exit' to quit): ").strip().lower()
        if topic.lower() == "exit":
            print("Exiting Kafka producer...")
            break

        if not topic:
            print("Topic name cannot be empty.")
            continue

        # Take message from user input
        message = input("Enter the message to send (or type 'exit' to quit): ")
        if message.lower() == "exit":
            print("Exiting Kafka producer...")
            break

        if not message:
            print("Message cannot be empty.")
            continue

        print(f"bootstrap.servers: {os.getenv('KAFKA_BOOTSTRAP_SERVERS', 'localhost:9092')}")

        # Ensure the topic exists or create it
        create_topic(topic)

        # Send the message to the specified topic
        producer.produce(topic, value=message.encode('utf-8'), callback=delivery_report)

        # Wait for all messages to be delivered
        producer.flush()
        print(f"Message '{message}' sent to topic '{topic}'")
    
    # Close the producer
    producer.close()

if __name__ == "__main__":
    main()

File: kafka-test/node/src/producer.js
const { Kafka, logLevel } = require("kafkajs");

// Kafka configuration
const kafkaConfig = {
  brokers: ["localhost:9092"],
};

console.log(`Kafka configuration: ${JSON.stringify(kafkaConfig)}`);
console.log(
  `bootstrap.servers: ${
    process.env.KAFKA_BOOTSTRAP_SERVERS || "localhost:9092"
  }`
);

const kafka = new Kafka({
  clientId: "kafka-producer-client",
  brokers: kafkaConfig.brokers,
  logLevel: logLevel.INFO,
});

const admin = kafka.admin();
const producer = kafka.producer();

async function createProducer() {
  await producer.connect();
  console.log("Kafka producer connected.");
  return producer;
}

async function deliveryReport(result) {
  // No direct callback like in confluent_kafka, but you can log success/failure here.
  if (result.errorCode) {
    console.error(`Message delivery failed: ${result.errorCode}`);
  } else {
    console.log(`Message delivered to ${result.topic} [${result.partition}]`);
  }
}

async function topicExists(topicName) {
  await admin.connect();
  const topics = await admin.listTopics();
  await admin.disconnect();
  return topics.includes(topicName);
}

async function createTopic(topicName) {
  const exists = await topicExists(topicName);
  if (!exists) {
    try {
      await admin.connect();
      await admin.createTopics({
        topics: [
          {
            topic: topicName,
            numPartitions: 1,
            replicationFactor: 1,
          },
        ],
      });
      console.log(`Topic '${topicName}' created successfully.`);
    } catch (e) {
      console.error(`Failed to create topic '${topicName}': ${e.message}`);
    } finally {
      await admin.disconnect();
    }
  } else {
    console.log(`Topic '${topicName}' already exists.`);
  }
}

async function main() {
  await createProducer();

  try {
    while (true) {
      // Take topic name from user input
      const topic = (
        await prompt("Enter the Kafka topic name (or type 'exit' to quit): ")
      ).trim();
      if (topic.toLowerCase() === "exit") {
        console.log("Exiting Kafka producer...");
        break;
      }

      if (!topic) {
        console.log("Topic name cannot be empty.");
        continue;
      }

      // Take message from user input
      const message = (
        await prompt("Enter the message to send (or type 'exit' to quit): ")
      ).trim();
      if (message.toLowerCase() === "exit") {
        console.log("Exiting Kafka producer...");
        break;
      }

      if (!message) {
        console.log("Message cannot be empty.");
        continue;
      }

      // Ensure the topic exists or create it
      await createTopic(topic);

      // Send the message to the specified topic
      await producer.send({
        topic: topic,
        messages: [{ value: message }],
      });

      console.log(`Message '${message}' sent to topic '${topic}'`);
    }
  } catch (error) {
    console.error(`An error occurred: ${error.message}`);
  } finally {
    await producer.disconnect();
    console.log("Kafka producer disconnected.");
  }
}

// Prompt wrapper for Node.js
async function prompt(question) {
  const readline = require("readline");
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

main().catch(console.error);
